README.md
# SwiftyDispatch

A lightweight GCD wrapper for Swift. Mostly copied from [MacRuby's Dispatch module](https://github.com/MacRuby/MacRuby/wiki/Dispatch::Queue-Class).

[![Version](https://img.shields.io/cocoapods/v/SwiftyDispatch.svg?style=flat)](http://cocoapods.org/pods/SwiftyDispatch)
[![License](https://img.shields.io/cocoapods/l/SwiftyDispatch.svg?style=flat)](http://cocoapods.org/pods/SwiftyDispatch)
[![Platform](https://img.shields.io/cocoapods/p/SwiftyDispatch.svg?style=flat)](http://cocoapods.org/pods/SwiftyDispatch)

## Usage

To run the example project, clone the repo, and run `pod install` from the Example directory first.

### Queue

#### Create queues

```swift
// Create a serial queue
let queue = Queue("Serial", .Serial)

// Create a concurrent queue
let queue = Queue("Concurrent", .Concurrent)

// It is serial by default
let queue = Queue("Concurrent")

// Get main queue
let queue = Queue.main()

// Get global queue via QOS
let queue = Queue.concurrent(.UserInteractive)

// Get global queue via priority
let queue = Queue.concurrent(priority: .Background)

```

#### Run blocks

```swift
// asynchronously
queue.async {
  data = true
}

// synchronously
queue.sync {
  data = true
}

```

#### Apply

```swift
queue.apply(10) {
  // some concurrent task
}
```

#### After

```swift
queue.apply(0.3) {
  // run after 0.3 seconds
}
```

### Semaphore

Semaphore provides an efficient mechanism to synchronizes threads via a combination of waiting and signaling.

This is especially useful for controlling access to limited resources.

```swift
let queue = Queue("sample")
let semaphore = Semaphore(0)
queue.after(1.0) {
  print "Hello"
  semaphore.signal
}
semaphore.wait // -> true
```


## Requirements

- Swift 2.0 (Xcode 7)

## Installation

### Cocoapods

**SwiftyDispatch** is available through [CocoaPods](http://cocoapods.org). To install
it, simply add the following line to your Podfile:

```ruby
pod "SwiftyDispatch"
```

### Carthage

You can use Carthage to install **SwiftyDispatch** by adding it to your Cartfile:

```ruby
github "siuying/SwiftyDispatch"
```

### Manually

To use this library in your project manually, just drag all swift files under ``Classes`` to the project tree.

## TODO?

- [x] Queue
- [x] Async/Sync
- [x] Apply
- [x] After
- [x] Semaphore
- [ ] Barrier
- [ ] Group
- [ ] Source
- [ ] Block class

## Author

Francis Chong, francis@ignition.hk

## License

Dispatch is available under the MIT license. See the LICENSE file for more info.


CHANGELOG.md
# 0.2.0

- Added Semaphore
- Fix After for not waiting correct time

# 0.1.0

Initial commit. Implemented:
- Queue
- Async/Sync
- Apply
- After

Example/Tests/Tests.swift
import UIKit
import XCTest
import Dispatch

class Tests: XCTestCase {
    
    override func setUp() {
        super.setUp()
        // Put setup code here. This method is called before the invocation of each test method in the class.
    }
    
    override func tearDown() {
        // Put teardown code here. This method is called after the invocation of each test method in the class.
        super.tearDown()
    }
    
    func testExample() {
        // This is an example of a functional test case.
        XCTAssert(true, "Pass")
    }
    
    func testPerformanceExample() {
        // This is an example of a performance test case.
        self.measureBlock() {
            // Put the code you want to measure the time of here.
        }
    }
    
}


Example/Tests/QueueTests.swift
import UIKit
import XCTest
import SwiftyDispatch
import Nimble

class QueueDispatchTests: XCTestCase {
    
    override func setUp() {
        super.setUp()
        // Put setup code here. This method is called before the invocation of each test method in the class.
    }
    
    override func tearDown() {
        // Put teardown code here. This method is called after the invocation of each test method in the class.
        super.tearDown()
    }
    
    func testCreate() {
        let queue = Queue("test", .Serial)
        expect(queue.label).to(equal("test"))
        
        let concurrent = Queue("concurrent", .Concurrent)
        expect(concurrent.label).to(equal("concurrent"))
        
        let defaultQueue = Queue("defaultQueue")
        expect(defaultQueue.label).to(equal("defaultQueue"))
        
    }

    func testMain() {
        let queue = Queue.main()
        expect(queue.label).to(equal("com.apple.main-thread"))        
    }
    
    func testConcurrent() {
        let queue = Queue.concurrent(.UserInitiated)
        expect(queue).toNot(beNil())
        expect(queue.label).to(contain("user-initiated"))
        
        let queue2 = Queue.concurrent(priority: .Background)
        expect(queue2).toNot(beNil())
        expect(queue2.label).to(contain("background"))
    }
    
    func testAsync() {
        var data : Bool = false
        let queue = Queue("test")
        queue.async { () -> Void in
            data = true
        }
        expect(data).toEventually(beTrue())
    }

    func testSync() {
        var data : Bool = false
        let queue = Queue("test")
        queue.sync { () -> Void in
            data = true
        }
        expect(data).toEventually(beTrue())
    }
    
    func testAfter() {
        let begin = NSDate()
        var data : Bool = false
        let queue = Queue.main()

        queue.after(0.1) {
            data = true
            expect(begin.timeIntervalSinceNow).to(beLessThanOrEqualTo(-0.1))
        }

        expect(data).toEventually(beTrue())
    }
    
    func testApply() {
        let queue = Queue("queue")
        var stack : [Int] = []

        queue.apply(5) { (i) -> Void in
            stack.append(i)
        }
        
        expect(stack.count).toEventually(equal(5))
        expect(stack).toEventually(equal([0,1,2,3,4]))
    }
    
}


Example/Tests/SemaphoreTests.swift
//
//  SemaphoreTests.swift
//  Dispatch
//
//  Created by Chan Fai Chong on 14/8/15.
//  Copyright © 2015 CocoaPods. All rights reserved.
//

import Foundation

import XCTest
import SwiftyDispatch
import Nimble

class SemaphoreDispatchTests: XCTestCase {
    func testCreate() {
        let queue = Queue("sample")
        let semaphore = Semaphore(0)

        var data = false
        let begin = NSDate()

        queue.after(0.1) { () -> Void in
            data = true
            semaphore.signal()
        }
        semaphore.wait()

        expect(data).to(beTrue())
        expect(begin.timeIntervalSinceNow).to(beLessThanOrEqualTo(-0.1))
    }

}


Classes/Semaphore.swift
//
//  Semaphore.swift
//  Pods
//
//  Created by Chan Fai Chong on 14/8/15.
//
//

import Foundation

public struct Semaphore {
    let semaphore : dispatch_semaphore_t

    ///
    /// Create a semaphore with given initial value, default 0
    /// - Parameter value: initial value of semaphore
    ///
    public init(_ value: Int = 0) {
        semaphore = dispatch_semaphore_create(value)
    }

    /// Wait (decrement) for a semaphore.
    ///
    /// Decrement the counting semaphore. If the resulting value is less than zero,
    /// this function waits for a signal to occur before returning.
    ///
    /// - Parameter time: Time to wait in second
    /// - Returns: True on success, or false if the timeout occurred.
    public func wait(time: NSTimeInterval) -> Bool {
        let nanoSeconds = Int64(Double(NSEC_PER_MSEC) * time)
        let delayTime = dispatch_time(DISPATCH_TIME_NOW, nanoSeconds)
        return dispatch_semaphore_wait(semaphore, delayTime) == 0
    }


    /// Wait (decrement) for a semaphore.
    ///
    /// Decrement the counting semaphore. If the resulting value is less than zero,
    /// this function waits for a signal to occur before returning.
    ///
    /// - Returns: True on success, or false if the timeout occurred.
    public func wait() -> Bool {
        return dispatch_semaphore_wait(semaphore, DISPATCH_TIME_FOREVER) == 0
    }
    
    public func signal() {
        dispatch_semaphore_signal(semaphore)
    }
}

Classes/SwiftyDispatch.swift
//
//  Dispatch.swift
//  BlueWhale2
//
//  Created by Chan Fai Chong on 14/8/15.
//  Copyright © 2015 Ignition Soft. All rights reserved.
//

import Foundation

public class Queue {
    public let queue : dispatch_queue_t
    
    public enum QueueType {
        case Serial
        case Concurrent
        var value : dispatch_queue_attr_t? {
            get {
                switch self {
                case .Serial:
                    return DISPATCH_QUEUE_SERIAL
                case .Concurrent:
                    return DISPATCH_QUEUE_CONCURRENT
                }
            }
        }
    }

    public enum Priority : Int {
        case Background
        case Default
        case High
        case Low

        var value : Int {
            get {
                switch self {
                case .Background:
                    return DISPATCH_QUEUE_PRIORITY_BACKGROUND
                case .Default:
                    return DISPATCH_QUEUE_PRIORITY_DEFAULT
                case .High:
                    return DISPATCH_QUEUE_PRIORITY_HIGH
                case .Low:
                    return DISPATCH_QUEUE_PRIORITY_LOW
                }
            }
        }
    }

    public enum QOS : UInt {
        case UserInteractive
        case UserInitiated
        case Default
        case Utility
        case Background

        var value : qos_class_t {
            get {
                switch self {
                case .UserInteractive:
                    return QOS_CLASS_USER_INTERACTIVE
                case .UserInitiated:
                    return QOS_CLASS_USER_INITIATED
                case .Default:
                    return QOS_CLASS_DEFAULT
                case .Utility:
                    return QOS_CLASS_UTILITY
                case .Background:
                    return QOS_CLASS_BACKGROUND
                }
            }
        }
    }

    public var label : String {
        get {
            return String.fromCString(dispatch_queue_get_label(queue))!
        }
    }

    public init(queue: dispatch_queue_t) {
        self.queue = queue
    }
    
    public init(_ label: String, _ type: QueueType = .Serial) {
        let labelBytes : UnsafePointer<Int8> = label.cStringUsingEncoding(NSUTF8StringEncoding)!.withUnsafeBufferPointer({
            $0.baseAddress
        })
        self.queue = dispatch_queue_create(labelBytes, type.value)
    }

    public func async(block: dispatch_block_t) {
        dispatch_async(self.queue, block)
    }
    
    public func sync(block: dispatch_block_t) {
        dispatch_sync(self.queue, block)
    }
    
    public func apply(iterations: Int, block: (Int) -> Void) {
        dispatch_apply(iterations, queue, block)
    }

    public func after(delayInSeconds: Double, block: dispatch_block_t) {
        let nanoSeconds = Int64(Double(NSEC_PER_MSEC) * delayInSeconds)
        let delayTime = dispatch_time(DISPATCH_TIME_NOW, nanoSeconds)
        dispatch_after(delayTime, queue, block)
    }
    
    public static func main() -> Queue {
        return Queue(queue: dispatch_get_main_queue())
    }
    
    public static func concurrent(qos: Queue.QOS) -> Queue {
        return Queue(queue: dispatch_get_global_queue(qos.value, 0))
    }

    public static func concurrent(priority priority: Queue.Priority) -> Queue {
        return Queue(queue: dispatch_get_global_queue(priority.value, 0))
    }
}

Classes/Queue.swift
//
//  Dispatch.swift
//  BlueWhale2
//
//  Created by Chan Fai Chong on 14/8/15.
//  Copyright © 2015 Ignition Soft. All rights reserved.
//

import Foundation

public struct Queue {
    public let queue : dispatch_queue_t
    
    public enum QueueType {
        case Serial
        case Concurrent
        var value : dispatch_queue_attr_t? {
            get {
                switch self {
                case .Serial:
                    return DISPATCH_QUEUE_SERIAL
                case .Concurrent:
                    return DISPATCH_QUEUE_CONCURRENT
                }
            }
        }
    }

    public enum Priority : Int {
        case Background
        case Default
        case High
        case Low

        var value : Int {
            get {
                switch self {
                case .Background:
                    return DISPATCH_QUEUE_PRIORITY_BACKGROUND
                case .Default:
                    return DISPATCH_QUEUE_PRIORITY_DEFAULT
                case .High:
                    return DISPATCH_QUEUE_PRIORITY_HIGH
                case .Low:
                    return DISPATCH_QUEUE_PRIORITY_LOW
                }
            }
        }
    }

    public enum QOS : UInt {
        case UserInteractive
        case UserInitiated
        case Default
        case Utility
        case Background

        var value : qos_class_t {
            get {
                switch self {
                case .UserInteractive:
                    return QOS_CLASS_USER_INTERACTIVE
                case .UserInitiated:
                    return QOS_CLASS_USER_INITIATED
                case .Default:
                    return QOS_CLASS_DEFAULT
                case .Utility:
                    return QOS_CLASS_UTILITY
                case .Background:
                    return QOS_CLASS_BACKGROUND
                }
            }
        }
    }

    public var label : String {
        get {
            return String.fromCString(dispatch_queue_get_label(queue))!
        }
    }

    /// Create a Queue object with a dispatch_queue_t
    ///
    /// - Parameter queue: underlying dispatch_queue_t object
    public init(queue: dispatch_queue_t) {
        self.queue = queue
    }

    /// Create a Queue with label and type
    ///
    /// - Parameter label: label of the queue
    /// - Parameter type: Type of the queue (QueueType)
    public init(_ label: String, _ type: QueueType = .Serial) {
        self.queue = dispatch_queue_create(label, type.value)
    }

    /// Run a block asynchronously
    ///
    /// - Parameter block: The block to run
    public func async(block: dispatch_block_t) {
        dispatch_async(self.queue, block)
    }

    /// Run a block synchronously
    ///
    /// - Parameter block: The block to run
    public func sync(block: dispatch_block_t) {
        dispatch_sync(self.queue, block)
    }

    /// Submits a block to run for multiple times.
    ///
    /// - Parameter iterations: Number of iterations
    /// - Parameter block: The block to run
    public func apply(iterations: Int, block: (Int) -> Void) {
        dispatch_apply(iterations, queue, block)
    }

    /// Schedule a block at a specified time.
    ///
    /// - Parameter delayInSeconds: Delay in number of seconds
    /// - Parameter block: The block to run
    public func after(delayInSeconds: Double, block: dispatch_block_t) {
        let nanoSeconds = Int64(Double(NSEC_PER_SEC) * delayInSeconds)
        let delayTime = dispatch_time(DISPATCH_TIME_NOW, nanoSeconds)
        dispatch_after(delayTime, queue, block)
    }

    /// Returns the default queue that is bound to the main thread.
    public static func main() -> Queue {
        return Queue(queue: dispatch_get_main_queue())
    }

    /// Returns a global queue by QOS
    /// - Parameter qos: the QOS of the queue
    public static func concurrent(qos: Queue.QOS) -> Queue {
        return Queue(queue: dispatch_get_global_queue(qos.value, 0))
    }

    /// Returns a global queue by Priotity
    /// - Parameter priority: the priority of the queue
    public static func concurrent(priority priority: Queue.Priority) -> Queue {
        return Queue(queue: dispatch_get_global_queue(priority.value, 0))
    }
}